src/gamepad/linux/linux-raw-gamepad-monitor.vala

```diff
@@ -0,0 +1,102 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.LinuxRawGamepadMonitor : Object, RawGamepadMonitor {
+	public delegate void RawGamepadCallback (RawGamepad raw_gamepad);
+
+	private static LinuxRawGamepadMonitor instance;
+
+	private GUdev.Client client;
+	private HashTable<string, RawGamepad> raw_gamepads;
+
+	private LinuxRawGamepadMonitor () {
+		client = new GUdev.Client ({"input"});
+		client.uevent.connect (handle_udev_client_callback);
+
+		// Initialize internally plugged in gamepads
+		client.query_by_subsystem ("input").foreach (initial_device_iterator);
+	}
+
+	public static LinuxRawGamepadMonitor get_instance () {
+		if (instance == null)
+			instance = new LinuxRawGamepadMonitor ();
+
+		return instance;
+	}
+
+	public void foreach_gamepad (RawGamepadCallback callback) {
+		raw_gamepads.foreach((identifier, raw_gamepad) => callback (raw_gamepad));
+	}
+
+	private void handle_udev_client_callback (string action, GUdev.Device device) {
+		if (device.get_device_file () == null)
+			return;
+
+		var identifier = device.get_device_file ();
+
+		if (!is_gamepad (device))
+			return;
+
+		switch (action) {
+		case "add":
+			var raw_gamepad = add_gamepad (device);
+			if (raw_gamepad != null)
+				gamepad_plugged (raw_gamepad);
+			break;
+		case "remove":
+			var raw_gamepad = remove_gamepad (device);
+			if (raw_gamepad != null) {
+				// This signal is emitted from here to simplify the code
+				raw_gamepad.unplugged ();
+			}
+			break;
+		}
+	}
+
+	private void initial_device_iterator (GUdev.Device device) {
+		if (device.get_device_file () == null)
+			return;
+
+		var identifier = device.get_device_file ();
+
+		if (!is_gamepad (device))
+			return;
+
+		add_gamepad (device);
+	}
+
+	private RawGamepad? add_gamepad (GUdev.Device device) {
+		var identifier = device.get_device_file ();
+		RawGamepad raw_gamepad;
+		try {
+			raw_gamepad = new LinuxRawGamepad (identifier);
+		} catch (FileError e) {
+			return null;
+		}
+
+		if (raw_gamepads.contains (identifier))
+			return null;
+
+		raw_gamepads.insert (identifier, raw_gamepad);
```

Comment:
This is the new style for hashtables. i had asked it earlier on irc and then we had confirmed to this. so i am not gonna edit this